### PR TITLE
[MOO-961] Auto update & patch known outdated dependencies

### PR DESCRIPTION
**Pull request by diego-antonelli**
## Checklist

-   Contains unit tests ✅ 
-   Contains breaking changes ❌
-   Compatible with: MX 9️⃣
-   Did you update version and changelog? ✅
-   PR title properly formatted (`[XX-000]: description`)? ✅

## This PR contains

-   [ ] Bug fix
-   [x] Feature
-   [ ] Refactor
-   [ ] Documentation
-   [ ] Other (describe)

## What is the purpose of this PR?
An option to warn the user about outdated dependencies and auto update them and install patches if necessary.

## Relevant changes
Auto migration with known outdated dependencies, including native ones

## What should be covered while testing?
Automation should be enough as MOO QA is gonna test

### CODE REVIEW

**Review by jeffersonRibeiro — APPROVED** [review]

File: packages/pluggable-widgets-tools/utils/migration.js [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+                    break;
+                default:
+                    return undefined;
+            }
+        })
+        .filter(Boolean);
+}
+
+function replaceOldDependency(listOfOutdatedDependency, target, message) {
+    if (listOfOutdatedDependency.length > 0) {
```
Comment (tamelbrun) [inline packages/pluggable-widgets-tools/utils/migration.js:103]: We should make this plural, `replaceOldDependencies` and `listOfOutdatedDependencies`.

File: packages/pluggable-widgets-tools/utils/migration.js [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    return new Promise(resolve =>
+        rl.question(yellow(question), answer => {
+            rl.close();
+            resolve(answer);
+        })
+    );
+}
+
+function getOutdatedDependencies(packageDependencies, listOfNewDependencies = dependencies) {
+    const createReturn = dep => ({
```
Comment (tamelbrun) [inline packages/pluggable-widgets-tools/utils/migration.js:66]: 🎄`versionInfo`?

File: packages/pluggable-widgets-tools/utils/migration.js [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        });
+    }
+}
+
+async function checkMigration() {
+    console.log("Checking if dependencies should be migrated...");
+    const packageJsonPath = join(process.cwd(), "package.json");
+    const packageJson = await readJson(packageJsonPath);
+    const args = process.argv;
+    if (args.indexOf("--skip-migration") === -1 && process.env.CI !== "true") {
```
Comment (tamelbrun) [inline packages/pluggable-widgets-tools/utils/migration.js:137]: 🎄`!includes` here as well.

**Review by tamelbrun — DISMISSED** [review]

File: packages/pluggable-widgets-tools/utils/migration.js [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+                case CheckType.MAJOR_MINOR:
+                    if (
+                        version[0] < extractVersions(dep.version)[0] ||
+                        (version[0] === extractVersions(dep.version)[0] && version[1] < extractVersions(dep.version)[1])
+                    ) {
+                        return createReturn(dep);
+                    }
+                    break;
+                default:
+                    return undefined;
```
Comment (tamelbrun) [inline packages/pluggable-widgets-tools/utils/migration.js:96]: This default case should never happen, maybe we can improve on this.

File: packages/pluggable-widgets-tools/utils/migration.js [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+const resolutionsOverrides = [
+    { name: "react", version: "18.2.0", check: CheckType.MAJOR_MINOR },
+    { name: "react-native", version: "0.70.7", check: CheckType.MINOR }
+];
+
+function extractVersions(version) {
+    return version
+        .replace(/^\D+/, "")
+        .split(".")
+        .map(stringValue => Number(stringValue));
```
[the remark this answers is not in the document]
Comment (jeffersonRibeiro) [reply packages/pluggable-widgets-tools/utils/migration.js]: Jelmer itself fixed ✨

File: packages/pluggable-widgets-tools/utils/migration.js [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+                case CheckType.MINOR:
+                    if (version[1] < extractVersions(dep.version)[1]) {
+                        return createReturn(dep);
+                    }
+                    break;
+                case CheckType.MAJOR_MINOR:
+                    if (
+                        version[0] < extractVersions(dep.version)[0] ||
+                        (version[0] === extractVersions(dep.version)[0] && version[1] < extractVersions(dep.version)[1])
+                    ) {
```
[the remark this answers is not in the document]
Comment (jeffersonRibeiro) [reply packages/pluggable-widgets-tools/utils/migration.js]: Jelmer itself fixed ✨

File: packages/pluggable-widgets-tools/utils/migration.js [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+                    addExtraDependencies(newPackageJson, "overrides", "The following overrides were added:");
+                    addExtraDependencies(newPackageJson, "resolutions", "The following resolutions were added:");
+
+                    // If any package requires a patch we make sure to install patch-package and add the script
+                    if (requirePatch) {
+                        if (!newPackageJson.devDependencies["patch-package"]) {
+                            newPackageJson.devDependencies["patch-package"] = "^6.5.0";
+                        }
+                        if (!newPackageJson.scripts.postinstall) {
+                            newPackageJson.scripts.postinstall = "patch-package";
```
[the remark this answers is not in the document]
Comment (jeffersonRibeiro) [reply packages/pluggable-widgets-tools/utils/migration.js]: Jelmer itself fixed ✨

File: packages/pluggable-widgets-tools/utils/migration.js [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+                        } else if (newPackageJson.scripts.postinstall.indexOf("patch-package") === -1) {
```
[the remark this answers is not in the document]
Comment (jeffersonRibeiro) [reply packages/pluggable-widgets-tools/utils/migration.js]: Jelmer itself fixed ✨

File: packages/pluggable-widgets-tools/utils/migration.js (hunk 3 above)
Comment (jeffersonRibeiro) [reply packages/pluggable-widgets-tools/utils/migration.js]: Jelmer itself fixed ✨

File: packages/pluggable-widgets-tools/utils/migration.js (hunk 4 above)
Comment (jeffersonRibeiro) [reply packages/pluggable-widgets-tools/utils/migration.js]: Jelmer itself fixed ✨

File: packages/pluggable-widgets-tools/utils/migration.js (hunk 1 above)
Comment (jeffersonRibeiro) [reply packages/pluggable-widgets-tools/utils/migration.js]: Jelmer itself fixed ✨

File: packages/pluggable-widgets-tools/utils/migration.js (hunk 2 above)
Comment (tamelbrun) [reply packages/pluggable-widgets-tools/utils/migration.js]: This was also refactored away :)

Comment (jeffersonRibeiro) [reply packages/pluggable-widgets-tools/utils/migration.js]: That one I couldn't verify, so I left it for later 😄

File: packages/pluggable-widgets-tools/CHANGELOG.md [hunk 9]
```diff
@@ -5,14 +5,19 @@ All notable changes to this tool will be documented in this file.
...
 
 ## [Unreleased]
+
 ### Added
 
 -   We added support to the typings generator for the data source caption (`{caption: string}`), introduced in version 9.24.0. This feature will give widget developers the ability to display the same caption for a data source, that Mendix uses in the Page Explorer, within a widget preview.
 
+-   We've added an option to automatically migrate known libraries to versions compatible with Studio Pro 9.24
+
+-   We add `@types/react`, `@types/react-dom` and `@types/react-native` as direct dependencies of the Pluggable Widgets Tools
```
Comment (mlaponder) [inline packages/pluggable-widgets-tools/CHANGELOG.md:15]: "We add" -> "We added" (that is the use construct in the rest of this file)

Comment (jeffersonRibeiro) [reply packages/pluggable-widgets-tools/CHANGELOG.md]: Fixed ✨

**Pull request merged**